internal/controller/pipelinerun_helper_test.go

```diff
@@ -0,0 +1,496 @@
+						CompletionTime: &metav1.Time{Time: time.Now().Add(5 * time.Minute)},
+					},
+					Status: v1.Status{
+						Conditions: v1.Conditions{
+							apis.Condition{
+								Message: "Tasks Completed: 3 (Failed: 0, Cancelled 0), Incomplete: 10, Skipped:1",
+								Reason:  "Running",
+								Status:  "Unknown",
+								Type:    apis.ConditionSucceeded,
+							},
+						},
+					},
+				},
+			}
+
+			err := k8sClient.Create(ctx, prPipelineRun)
+			Expect(err).ToNot(HaveOccurred())
+			Eventually(func() bool {
+				err = k8sClient.Get(ctx, prPipelineLookupKey, testPipelineRun)
+				return err == nil
+			}, timeout, interval).Should(BeTrue())
+
+			// Finalizer and annotation should not exists
+			Expect(controllerutil.ContainsFinalizer(testPipelineRun, NotificationPipelineRunFinalizer)).To(BeFalse())
+			Expect(metadata.HasAnnotationWithValue(testPipelineRun, NotificationPipelineRunAnnotation, NotificationPipelineRunAnnotationValue)).To(BeFalse())
+		})
+		Context("Pipelinerun without finalizer and annotation", func() {
+			It("AddFinalizerToPipelineRun should succeed", func() {
+				err := AddFinalizerToPipelineRun(ctx, testPipelineRun, nsr, NotificationPipelineRunFinalizer)
+				Expect(err).ToNot(HaveOccurred())
+				Eventually(func() bool {
+					err := k8sClient.Get(ctx, prPipelineLookupKey, testPipelineRun)
+					Expect(err).ToNot(HaveOccurred())
+					return controllerutil.ContainsFinalizer(testPipelineRun, NotificationPipelineRunFinalizer)
+				}, timeout, interval).Should(BeTrue())
+			})
+			It("AddAnnotationToPipelineRun should succeed", func() {
+				err := AddAnnotationToPipelineRun(ctx, testPipelineRun, nsr, NotificationPipelineRunAnnotation, NotificationPipelineRunAnnotationValue)
+				Expect(err).ToNot(HaveOccurred())
+				Eventually(func() bool {
+					err := k8sClient.Get(ctx, prPipelineLookupKey, testPipelineRun)
+					Expect(err).ToNot(HaveOccurred())
+					return metadata.HasAnnotationWithValue(testPipelineRun, NotificationPipelineRunAnnotation, NotificationPipelineRunAnnotationValue)
+				}, timeout, interval).Should(BeTrue())
+			})
+			It("RemoveFinalizerFromPipelineRun should succeed if finalizer does not exist", func() {
+				err := RemoveFinalizerFromPipelineRun(ctx, testPipelineRun, nsr, NotificationPipelineRunFinalizer)
+				Expect(err).ToNot(HaveOccurred())
+				Eventually(func() bool {
+					err := k8sClient.Get(ctx, prPipelineLookupKey, testPipelineRun)
+					Expect(err).ToNot(HaveOccurred())
+					return !controllerutil.ContainsFinalizer(testPipelineRun, NotificationPipelineRunFinalizer)
+				}, timeout, interval).Should(BeTrue())
+			})
+		})
+		AfterEach(func() {
+			err := k8sClient.Delete(ctx, testPipelineRun)
+			Expect(err == nil || errors.IsNotFound(err)).To(BeTrue())
+		})
+	})
+	Describe("Successfully adding/removing resources when exist", func() {
+		BeforeEach(func() {
+			// Create a pull_request pipelinerun
+			prPipelineRun = &tektonv1.PipelineRun{
```

Comment:
type-name matching

Reply:
In line 24:
`prPipelineRun *tektonv1.PipelineRun`